kubernetes-deployment/maintainlib/etcdfix.py

```diff
@@ -0,0 +1,236 @@
+
+        self.prepare_package(bad_node_config, "etcd-reconfiguration-restart")
+
+        script_package = "etcd-reconfiguration-restart.tar"
+        src_local = "parcel-center/{0}".format(bad_node_config["nodename"])
+        dst_remote = "/home/{0}".format(bad_node_config["username"])
+
+        if common.sftp_paramiko(src_local, dst_remote, script_package, bad_node_config) == False:
+            return
+
+        commandline = "tar -xvf {0}.tar && sudo ./{0}/{1}.sh".format("etcd-reconfiguration-restart", "restart-etcd-server")
+
+        if common.ssh_shell_paramiko(bad_node_config, commandline) == False:
+            return
+
+        self.logger.info("Successfully restarting bad etcd server on node {0}".format(bad_node_config["nodename"]))
+
+        if self.clean_flag:
+            self.delete_packege(bad_node_config)
+
+
+
+    def get_etcd_leader_node(self):
+
+        # Get leader node.
+        host_list = list()
+
+        for host in self.cluster_config['mastermachinelist']:
+            host_list.append((self.cluster_config['mastermachinelist'][host]['hostip'], 4001))
+
+        client = etcd.Client(host=tuple(host_list), allow_reconnect=True)
+
+        etcdid = client.leader['name']
+
+        for host in self.cluster_config['mastermachinelist']:
+            if etcdid == self.cluster_config['mastermachinelist'][host]['etcdid']:
+                self.logger.debug("Current leader of etcd-cluster: {0}".format(self.cluster_config['mastermachinelist'][host]))
+                return self.cluster_config['mastermachinelist'][host]
+
+        self.logger.error("Can't find the leader of etcd.")
+        return None
+
+
+
+    def get_etcd_peer_ip_list(self, bad_node_config):
+
+        etcd_cluster_ips_peer = ""
+        separated = ""
+
+        host_list = list()
+
+        for host in self.cluster_config['mastermachinelist']:
+            host_list.append((self.cluster_config['mastermachinelist'][host]['hostip'], 4001))
+
+        client = etcd.Client(host=tuple(host_list), allow_reconnect=True)
+
+        member_dict = client.members
+        for member_hash in member_dict:
+
+            etcd_id = member_dict[member_hash]['name']
+            peer_url = member_dict[member_hash]['peerURLs'][0]
+
+            if etcd_id == "":
+                # new member before announcing, etcdid will be empty.
+                continue
+
+            ip_peer = "{0}={1}".format(etcd_id, peer_url)
+
+            etcd_cluster_ips_peer = etcd_cluster_ips_peer + separated + ip_peer
+
+            separated = ","
+
+        new_etcd_id = bad_node_config['etcdid']
+        peer_url = bad_node_config['hostip']
+        ip_peer = "{0}=http://{1}:2380".format(new_etcd_id, peer_url)
+        etcd_cluster_ips_peer = etcd_cluster_ips_peer + separated + ip_peer
+
+        self.logger.debug("New etcd-initial-cluster: {0}".format(etcd_cluster_ips_peer))
+
+        return etcd_cluster_ips_peer
+
+
+
+    def run(self):
```

Comment:
do you want to add a repair-infra-node.sh to demonstrate how to use the class?

Reply:
Later